syntax = "proto3";

package csw_protobuf;

import "csw_protobuf/units.proto";
import "csw_protobuf/keytype.proto";
import "scalapb/scalapb.proto";
import "csw_protobuf/parameter_types.proto";

option (scalapb.options) = {
    import: "csw.event.client.pb._"
    import: "csw.event.client.pb.TypeMapperSupport._"
    import: "csw.params.core.models._"
    single_file: true
};

message PbParameter {
    string name = 1;
    PbUnits units = 2;
    PbKeyType keyType = 3;
    oneof items {
        ChoiceItems choiceItems = 50;
        RaDecItems raDecItems = 51;
        StructItems structItems = 52;
        InstantItems instantItems = 53;

        StringItems stringItems = 54;
        BooleanItems booleanItems = 55;
        CharItems charItems = 56;

        ByteItems byteItems = 57;
        ShortItems shortItems = 58;
        IntItems intItems = 59;
        LongItems longItems = 60;
        DoubleItems doubleItems = 61;
        FloatItems floatItems = 62;

        ByteArrayItems byteArrayItems = 63;
        ShortArrayItems shortArrayItems = 64;
        IntArrayItems intArrayItems = 65;
        LongArrayItems longArrayItems = 66;
        FloatArrayItems floatArrayItems = 67;
        DoubleArrayItems doubleArrayItems = 68;

        ByteMatrixItems byteMatrixItems = 69;
        ShortMatrixItems shortMatrixItems = 70;
        IntMatrixItems intMatrixItems = 71;
        LongMatrixItems longMatrixItems = 72;
        FloatMatrixItems floatMatrixItems = 73;
        DoubleMatrixItems doubleMatrixItems = 74;
    }
}

message PbStruct {
    repeated PbParameter paramSet = 1;
}

message StructItems {
    option (scalapb.message).extends = "ItemType[Struct]";
    option (scalapb.message).companion_extends = "ItemTypeCompanion[StructItems]";
    repeated PbStruct values = 1 [(scalapb.field).type = "Struct"];
}
